compiler/hash-semantics/src/passes/discovery/utils.rs

```diff
@@ -0,0 +1,133 @@
+                return Ok(Node::at(ty, NodeOrigin::Given(repr_hint.origin)));
+            }
+        }
+
+        // Otherwise we have to loop through all of the variants, and check if they have
+        // a direct discriminant value...
+        let mut prev_discr = None;
+        let mut has_discriminant_attr = false;
+        let mut is_signed = false;
+
+        for variant in def.entries.iter() {
+            if let Some(discr_annot) = attr_store().get_attr(variant.id(), attrs::DISCRIMINANT) {
+                let discr = discr_annot.get_arg(0).unwrap().as_int().big_value();
+                has_discriminant_attr = true;
+                is_signed |= discr < BigInt::from(0);
+
+                // Otherwise, we set the previous discriminant to the current one.
+                prev_discr = Some(discr);
+            } else if let Some(prev) = prev_discr {
+                // We just add one to the the biggest discriminant we have seen so far.
+                prev_discr = Some(prev + 1);
+            } else {
+                // Initialise it with a zero if no initial discriminant is given.
+                prev_discr = Some(BigInt::from(0));
+            }
+        }
+
+        // If we get no variants, we still assume that the discriminant will use the
+        // smallest unsigned integer.
+        let max_discr = prev_discr.unwrap_or_default();
+        let min_bits = max_discr.bits();
+
+        // If we didn't have any registered discriminants, then we assume that the
+        // type of the discriminants is the smallest unsigned integer that can represent
+        // the number of variants.
+        //
+        // @@NonExhaustive: if we add a `non_exhaustive` attribute, what is the type
+        // that we assume now, is it perhaps an i32 or isize?
+        if !has_discriminant_attr || min_bits == 0 {
+            let int = Integer::fit_unsigned(def.entries.len() as u128);
+            return Ok(Node::at(IntTy::from_integer(int, false), NodeOrigin::Generated));
+        }
+
+        // Check whether the discriminant can fit in a `u8`, `u16`, `u32`, `u64`,
+        // `u128`.
+        let Some(int) = Integer::from_size(Size::from_bits(min_bits)) else {
+            panic!("error: discriminant too large!")
+        };
+
+        Ok(Node::at(IntTy::from_integer(int, is_signed), NodeOrigin::Generated))
+    }
+
+    /// Check if any of the discriminants that are assigned to a variant are
+    /// duplicated across the enum definition (whether by overflow or by
+    /// explicit means).
+    pub(super) fn check_enum_discriminants(
+        &self,
+        discrs: &[Node<Discriminant>],
+        _discr_ty: Node<IntTy>,
+    ) -> SemanticResult<()> {
+        let mut i = 0;
+
+        // Check if any of the discriminants have been used more than once.
+        while i < discrs.len() {
+            // @@Todo: maybe make this an error collector of sorts?
+
+            let mut o = i + 1;
+            while o < discrs.len() {
```

Comment:
can be replaced by `for (a, b) in discrs.combinations(2)` from `itertools`?

Reply:
Actually, its not that simple:
- Firstly, you cannot do `(a, b)` since `combinations(2)` wiill return a `Vec<...>`
- The code will likely be messier because getting the elements from the vec will need to do something like:
   ```rs
   let [a, b] = pair else { unreachable!() };
    ```
- Its allocating a new vec for each pair, seems unnecessary

Reply:
Damn that's sad, I thought it would return an iterator... this is why one needs dependent types

Reply:
Indeed